libraries/AP_ESC_Telem/AP_ESC_Telem.cpp

```diff
@@ -35,23 +39,327 @@ AP_ESC_Telem::AP_ESC_Telem()
+        _telem_data[esc_index].temperature_cdeg = new_data.temperature_cdeg;
+    }
+    if (data_mask & AP_ESC_Telem_Backend::TelemetryType::MOTOR_TEMPERATURE) {
+        _telem_data[esc_index].motor_temp_cdeg = new_data.motor_temp_cdeg;
+    }
+    if (data_mask & AP_ESC_Telem_Backend::TelemetryType::VOLTAGE) {
+        _telem_data[esc_index].voltage_cv = new_data.voltage_cv;
+    }
+    if (data_mask & AP_ESC_Telem_Backend::TelemetryType::CURRENT) {
+        _telem_data[esc_index].current_ca = new_data.current_ca;
+    }
+    if (data_mask & AP_ESC_Telem_Backend::TelemetryType::CONSUMPTION) {
+        _telem_data[esc_index].consumption_mah = new_data.consumption_mah;
+    }
+    if (data_mask & AP_ESC_Telem_Backend::TelemetryType::USAGE) {
+        _telem_data[esc_index].usage_s = new_data.usage_s;
+    }
+
+    _telem_data[esc_index].count++;
+    _telem_data[esc_index].types |= data_mask;
+    _telem_data[esc_index].last_update_ms = AP_HAL::millis();
+}
+
+// record an update to the RPM together with timestamp, this allows the notch values to be slewed
+// this should be called by backends when new telemetry values are available
+void AP_ESC_Telem::update_rpm(uint8_t esc_index, uint16_t new_rpm, float error_rate)
+{
+    if (esc_index > ESC_TELEM_MAX_ESCS) {
+        return;
+    }
+    uint32_t now = AP_HAL::micros();
+    volatile AP_ESC_Telem_Backend::RpmData& rpmdata = _rpm_data[esc_index];
+
+    rpmdata.prev_rpm = rpmdata.rpm;
+    rpmdata.rpm = new_rpm;
+    rpmdata.update_rate_hz = 1.0e6f / (now - rpmdata.last_update_us);
+    rpmdata.last_update_us = now;
+
+#ifdef ESC_TELEM_DEBUG
+    hal.console->printf("RPM: rate=%.1fhz, rpm=%d)\n", rpmdata.update_rate_hz, new_rpm);
+#endif
+}
+
+void AP_ESC_Telem::init(void)
 {
+    if (_initialised) {
+        return;
+    }
+
 #if HAL_MAX_CAN_PROTOCOL_DRIVERS
     const uint8_t num_drivers = AP::can().get_num_drivers();
     for (uint8_t i = 0; i < num_drivers; i++) {
         if (AP::can().get_driver_type(i) == AP_CANManager::Driver_Type_ToshibaCAN) {
             AP_ToshibaCAN *tcan = AP_ToshibaCAN::get_tcan(i);
             if (tcan != nullptr) {
-                usage_sec = tcan->get_usage_seconds(esc_id);
-                return true;
+                add_backend(tcan);
+                break;
             }
-            return true;
         }
     }
 #endif
-    return false;
+
+    _initialised = true;
+}
+
+// log ESC telemetry at 10Hz
+void AP_ESC_Telem::update()
+{
+    AP_Logger *logger = AP_Logger::get_singleton();
+
+    // Push received telemtry data into the logging system
+    if (logger && logger->logging_enabled()) {
+
+        for (uint8_t i = 0; i < ESC_TELEM_MAX_ESCS; i++) {
+            if (_telem_data[i].last_update_ms > _last_telem_log_ms[i]
+                || _rpm_data[i].last_update_us > _last_telem_log_ms[i] * 1000) {
+
+                float rpm = 0.0f;
+                get_rpm(i, rpm);
+
+                logger->Write_ESC(i, AP_HAL::micros64(),
```

Comment:
change log format for current and voltage to float

Reply:
Fixed